modules/core/src/matmul.dispatch.cpp

```diff
@@ -59,6 +60,138 @@ namespace cv
+    }
+    auto &device = ocl::Device::getDefault();
+    bool SupportFP64 = device.hasFP64(),
+         SupportFP16 = device.hasFP16();
+    if (!SupportFP64 && depth == CV_64F) {
+        return false;
+    }
+    if (!SupportFP16 && depth == CV_16F) {
+        return false;
+    }
+
+    Size sizeA = matA.size(), sizeB = matB.size(), sizeC = haveC ? matC.size() : Size(0, 0);
+    bool transA_ = (flags & GEMM_1_T) != 0, transB_ = (flags & GEMM_2_T) != 0, transC_ = (flags & GEMM_3_T) != 0;
+
+    if (transA_) {
+        sizeA = Size(sizeA.height, sizeA.width);
+    }
+    if (transB_) {
+        sizeB = Size(sizeB.height, sizeB.width);
+    }
+    if (haveC && transC_) {
+        sizeC = Size(sizeC.height, sizeC.width);
+    }
+
+    Size sizeD(sizeB.width, sizeA.height);
+
+    CV_CheckEQ(sizeA.width, sizeB.height, "Invalid dimension for matrix multiplification");
+    if (haveC) { // TODO: support matC broadcasting
+        CV_CheckTrue(sizeC == sizeD, "Shape of matC is not equal to the shape of matD");
+    }
+
+    matD.create(sizeD, type);
+    if (matA.offset() % esz != 0 || matA.step() % esz != 0 ||
+        matB.offset() % esz != 0 || matB.step() % esz != 0 ||
+        (haveC && (matC.offset() % esz != 0 || matC.step() % esz != 0)) )
+        return false;
+
+    UMat A = matA.getUMat(), B = matB.getUMat(), D = matD.getUMat();
+    if (!ocl::internal::isCLBuffer(A) || !ocl::internal::isCLBuffer(B) || !ocl::internal::isCLBuffer(D)) {
+        return false;
+    }
+    if (haveC) {
+        UMat C = matC.getUMat();
+        if (!ocl::internal::isCLBuffer(C))
+            return false;
+
+        if (transC_) {
+            transpose(matC, D);
+        } else {
+            matC.copyTo(D);
+        }
+    } else {
+        D.setTo(Scalar::all(0));
+    }
+
+    int M = sizeD.height, N = sizeD.width, K = sizeA.width;
+    int lda = static_cast<int>(A.step / esz),
+        ldb = static_cast<int>(B.step / esz),
+        ldc = static_cast<int>(D.step / esz);
+    int offsetA = static_cast<int>(A.offset / esz),
+        offsetB = static_cast<int>(B.offset / esz),
+        offsetC = static_cast<int>(D.offset / esz);
+
+    cl_command_queue queue = (cl_command_queue)ocl::Queue::getDefault().ptr();
+    CLBlastTranspose transA = transA_ ? CLBlastTransposeYes : CLBlastTransposeNo,
+                     transB = transB_ ? CLBlastTransposeYes : CLBlastTransposeNo;
+    CLBlastLayout layout = CLBlastLayoutRowMajor;
+    CLBlastStatusCode status = CLBlastUnknownError;
+
+    if (type == CV_32FC1) {
+        status = CLBlastSgemm(layout, transA, transB, M, N, K,
+                              (float)alpha,
+                              (const cl_mem)A.handle(ACCESS_READ), offsetA, lda,
+                              (const cl_mem)B.handle(ACCESS_READ), offsetB, ldb,
+                              (float)beta,
+                              (cl_mem)D.handle(ACCESS_RW), offsetC, ldc,
+                              &queue, NULL);
```

Comment:
What is about async processing?

Reply:
Do you mean async processing several calls to Sgemm for example?